tango/integrations/transformers/prefix_transformer.py

```diff
@@ -0,0 +1,111 @@
+import random
+from typing import Dict, Any
+import logging
+
+import torch
+from torch import nn
+from transformers import PreTrainedModel
+from transformers.modeling_outputs import CausalLMOutputWithCrossAttentions
+
+from tango.integrations.torch import Model
+
+
+logger = logging.getLogger(__name__)
+
+
+class WithPrefixEmbedding(nn.Module):
+    def __init__(
+        self, original_embedding: nn.Embedding, prefix_length: int, random_seed: int = 1940
+    ):
+        super().__init__()
+
+        self.prefix_length = prefix_length
+        self.original_embedding = original_embedding
+        self.prefix_embedding = nn.Embedding(prefix_length, self.original_embedding.embedding_dim)
+
+        # following Lester et al. 2021 in initializing using the top 5000 random vocabs
+        r = random.Random(random_seed)
+        indices = r.sample(range(5000), prefix_length)
+        with torch.no_grad():
+            self.prefix_embedding.weight.copy_(self.original_embedding.weight[indices])
+
+    def forward(self, input: torch.Tensor):
+        embedded_prefix = self.prefix_embedding(input[:, : self.prefix_length])
+        embedded_rest = self.original_embedding(input[:, self.prefix_length :])
+        return torch.cat([embedded_prefix, embedded_rest], dim=1)
+
+
+def make_prefix_transformer(model: Model, prefix_length: int) -> Model:
+    assert isinstance(model, PreTrainedModel)
+
+    model.set_input_embeddings(WithPrefixEmbedding(model.get_input_embeddings(), prefix_length))
+
+    def patch_tensor(kwargs: Dict[str, torch.Tensor], key: str, value: Any = 0) -> None:
+        try:
+            t = kwargs[key]
+        except KeyError:
+            return
+        zero_prefix = torch.full(
```

Comment:
`full_like`?

Reply:
Shape is different

Reply:
`new_full` then

Reply:
Though I don't particularly like the un-readability of `new_full`

Reply:
Ah, `new_full` is indeed better. Readability concerns aside, I like that when I use a method like that, it'll just work when the torch devs think of a new piece of metadata that could be attached to a tensor.

Reply:
0cc761f